md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2007 @@
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries each containing
+    - A [control-flow type](#control-flow-type).
+    - A *limit* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During validation, if either stack is empty when an element is to be popped from
+it, or an element is accessed by index less than zero or not less than the
+stack's length, validation fails. If a value is popped from the type stack such
+that the length of the type stack becomes less than the control-flow stack top's
+limit value, validation fails.
+
+#### Control Flow Type
+
+A *control-flow type* is either a [type](#types), `any`, which satisfies any
+type check, or `void`, which satisfies no check.
+
+> `any` is effectively serving as a kind of bottom type, and `void` is
+effectively serving as a kind of top type. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Control Flow Type Merge
+
+To merge two [control-flow types]:
+ - If either control-flow type is `any`, the result is the other control flow
+   type.
+ - Otherwise, if the control-flow types are the same, the result is that control
+   flow type.
+ - Otherwise, validation fails.
+
+#### Function Validation Initialization
+
+If the instruction sequence is empty, or if the last instruction in the sequence
+is not an `end`, validation fails.
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's control-flow type is
+the first return type of the function, if it has any, or `void` otherwise. Its
+limit value is zero.
+
+> The final `end` instruction may be implicit in some representations.
+
+#### Function Body Validation
+
+The instruction at the current position is remembered, and the current position
+is incremented to point to the position following it. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, [generic validation](#generic-instruction-validation) is
+performed.
+
+If the current position is now past the end of the sequence,
+[function return validation](#function-return-validation) is initiated and
+validation of the function is thereafter complete.
+
+Otherwise, [validation](#function-body-validation) is resumed.
+
+TODO: `unreachable`, `br`, `return`, and `br_table` return type validation.
+
+#### Generic Instruction Validation
+
+For each operand [type] in the instruction's signature in reverse order, a
+type is popped from the type stack and checked to match it. Each return type of
+the instruction's signature is then pushed onto the type stack.
+
+Then, if the instruction's signature has no return types, and the length of the
+type stack is greater than the control-flow stack top's limit value, validation
+fails.
+
```

Comment:
@sunfishcode Yes, thank you, I see the `pop` validate check now, but am still lost on what the purpose of the above clause is? If the pop's are already validated then how could the above then fail validation, and what does it have to do with a lack of results for an instruction?